### PR TITLE
Fix a race in CachingDictionary.

### LINKED ISSUE #60259: Intermittent build failures with  "error CS0104: `X` is an ambiguous reference between `System.X` and `System.X`"
**Version Used**: 
- MSBuild version = "17.1.0+ae57d105c"
- dotnet sdk 6.0.200

**Steps to Reproduce**:

1. Clone https://github.com/duncanp-sonar/BugRepro_AmbiguousTypesNet6
2. Repro steps in the repo README _(TL;DR: do a sonar analysis when running the build, which injects sonar Roslyn analyzers; run the build multiple times until it fails; it intermittently fails - only ~20% of the runs)_

**Expected Behavior**: build & analysis is successfull

**Actual Behavior**:
Build fails intermittently with "ambiguous reference" errors e.g. 

>(CoreCompile target) -> d:\support\S4MSB\AmbiguousTypesNet6\Class2.cs(13,32): error CS0104: 'TimeSpan' is an ambiguous reference between 'System.TimeSpan' and 'System.TimeSpan' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj] d:\support\S4MSB\AmbiguousTypesNet6\Entities.cs(14,16): error CS0104: 'DateTime' is an ambiguous reference between 'System.DateTime' and 'System.DateTime' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj] d:\support\S4MSB\AmbiguousTypesNet6\Entities.cs(19,16): error CS0104: 'DateTime' is an ambiguous reference between 'System.DateTime' and 'System.DateTime' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj]

**Binary Logs**:

You can check the binary logs for a successful build and a failed one: https://github.com/duncanp-sonar/BugRepro_AmbiguousTypesNet6/tree/master/logs. 

Note: the `csc.exe` invocations are identical.

**Details**:

Initially reported on [sonar community](https://community.sonarsource.com/t/timespan-is-an-ambiguous-reference-between-system-timespan-and-system-timespan/59245/9).  Some users have reported other numbers for their projects (only 1 / 10 of the builds succeed). A user reported that by removing `ImplicitUsings`, the problem would stop manifesting itself.

Separately, this has been reported [on StackOverflow as well](https://stackoverflow.com/questions/71171933/error-cs0104-guid-is-an-ambiguous-reference-between-system-guid-and-system) (seems not to be related to the sonar Roslyn analyzers).

Can this be related to https://github.com/dotnet/roslyn/issues/59564 ?

Could you please help:
- Are you aware of non-deterministic behavior inside the compiler that could lead to the stochastic errors we are witnessing?
- Do you have any hints on what to investigate in order to drill down to the root cause of the issue? 

Thank you!

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #60259.

### CODE REVIEW

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Comment by AlekseyTs** [comment]
@RikkiGibson, @dotnet/roslyn-compiler For the second review.

File: src/Compilers/Core/Portable/Collections/CachingDictionary.cs [hunk 1]
```diff
@@ -237,7 +237,7 @@ private static bool IsNotFullyPopulatedMap([NotNullWhen(returnValue: false)] IDi
 
                     if (!existingMap.TryGetValue(key, out elements))
                     {
-                        elements = _getElementsOfKey(key);
+                        elements = existingMap.GetOrAdd(key, _getElementsOfKey(key));
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Collections/CachingDictionary.cs:240]: It looks like GetOrAdd has an overload which could take `_getElementsOfKey` and `key` and only call the delegate if it is necessary

**Comment by AlekseyTs** [comment]
Sorry, I have to rebase the branch in order to target release/dev17.3

**Review by cston — APPROVED** [review]

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**